Add blog post about CQRS design pattern (#1184)
debezium/debezium.github.io: _posts/2025-10-14-cqrs.adoc

== CQRS with Database Built-in Replication

As mentioned earlier, the simplest implementation of the CQRS pattern is when your database provides built-in replication capabilities.
In our example, we use the PostgreSQL database, which supports link:https://www.postgresql.org/docs/current/runtime-config-replication.html[streaming replication].
Schematically, our demo application looks like this:

++++
<div class="imageblock centered-image">
    <img src="/assets/images/2025-10-14-cqrs/cqrs-replication.png" class="responsive-image" alt="Demo application with DB replication">
</div>
++++

To enable streaming replication between two PostgreSQL instances, we first need to create a publication on the primary instance and a replication role with `SELECT` rights on the required tables:

[source,sql,options="nowrap"]
----
CREATE PUBLICATION poll_pub FOR ALL TABLES;
CREATE ROLE replicator WITH REPLICATION LOGIN PASSWORD 'replicator';
GRANT SELECT ON ALL TABLES IN SCHEMA public TO replicator;
----

For simplicity, we granted `SELECT` rights on all tables.
In a production setup, you should limit these privileges to only the tables that actually need to be replicated.

> Because Hibernate may drop and recreate tables as part of schema updates, we also need to replicate `DELETE` operations.

Reviewer: See below - dropping tables does not emit `DELETE` ops

    PUBLICATION poll_pub;
----

After setting this up, all changes will automatically be replicated from the primary database, which serves the writer microservice, to the replica database used by the reader microservice.
This is a straightforward and reliable solution that does not require any third-party tools.
It can serve well until you need to integrate multiple data sources or use a different database type for the reader microservice.

== CQRS in Heterogeneous Environments

In more heterogeneous environments, where different database technologies are used for the write and read sides, built-in replication mechanisms are usually not sufficient.
You may want to replicate data between different types of databases or aggregate data from several systems into a single read-optimized store.
In such cases, a Change Data Capture (CDC) solution like Debezium becomes extremely useful.

Debezium continuously monitors changes in a source database by reading its transaction log and publishes these changes as events to Apache Kafka.
From there, the data can be written into another database or system using a suitable Kafka Connect sink connector.

> For database which support inserting data over JDBC, we can use link:https://debezium.io/documentation//reference/connectors/jdbc.html[JDBC sink connectr], eventually for MongoDB we can use link:https://debezium.io/documentation//reference/connectors/mongodb-sink.html[MongoDB sink connector].

Reviewer: ```suggestion
For database which support inserting data over JDBC, we can use link:https://debezium.io/documentation//reference/connectors/jdbc.html[JDBC sink connector], eventually for MongoDB we can use link:https://debezium.io/documentation//reference/connectors/mongodb-sink.html[MongoDB sink connector].
```

For databases which don't support inserts over JDBC, other than MongoDB, we have to rely on the custom sink connector provided by the database vendor.

However, Debezium usage is not limited to Kafka Connect runtime.

> We can e.g. embed Debezium into our custom microservice, but in such case we have to implement sink part (part which writes the captured data into target database) ourselves.

Reviewer: Link to Debezium engine and Debezium extension would be good in this sentence.

The following diagram illustrates this setup:

++++
<div class="imageblock centered-image">
	<img src="/assets/images/2025-10-14-cqrs/cqrs-debezium.png" class="responsive-image" alt="Demo application with Debezium-based replication">
</div>
++++

For simplicity, in out demo application we will also include Kafka Connect as the Debezium runtime.
We will show two examples of using Debezium for data replication:

* Replicating data from PostgreSQL to MySQL using the Debezium JDBC Sink Connector.
* Replicating data from PostgreSQL to QuestDB using the QuestDB Kafka Sink Connector.

=== Replicating from PostgreSQL to MySQL

> This setup demonstrates how to use Debezium with Kafka Connect to replicate data between two relational databases and target database supports JDBC API.

Reviewer: ```suggestion
This setup demonstrates how to use Debezium with Kafka Connect to replicate data between two relational databases and the target database supports JDBC API.
```

    }
}
----

This configuration captures all changes in the specified tables and publishes them to Kafka topics under the prefix `polls`.

On the sink side, we use the Debezium JDBC Sink Connector to insert these records into a MySQL database.
Since the JDBC sink understands Debezium’s event format, no single message transformation (SMT) is required.

[source,json,options="nowrap"]
----
{
    "name": "mysql-sink",
    "config": {
        "connector.class": "io.debezium.connector.jdbc.JdbcSinkConnector",
        "tasks.max": "1",
        "connection.url": "jdbc:mysql://mysql:3306/polls",
        "connection.username": "mysqluser",
        "connection.password": "mysqlpw",
        "insert.mode": "upsert",
        "delete.enabled": "true",
        "primary.key.mode": "record_key",
        "schema.evolution": "basic",
        "collection.name.format": "${source.table}",
        "topics": "polls.public.voteentity, polls.public.optionentity, polls.public.pollentity, polls.public.pollentity_optionentity"
    }
}
----

In this configuration, the `insert.mode` is set to `upsert`, which ensures that both new inserts and updates are properly reflected in the target database.

> Deletes are enabled as well, since Hibernate may drop tables.

Reviewer: I belive dropping tables is unrelated to this setting. Also drop table does not emit deletes for recors so the reference is probably not correct.

Author: yes, you are right, it doesn't drop the table, it does ordinary deletes, e.g.:
```
writer-1    | [Hibernate] 
writer-1    |     delete 
writer-1    |     from
writer-1    |         PollEntity_VoteEntity 
writer-1    |     where
writer-1    |         PollEntity_id=?
```

Conveniently Debezium provides link:https://debezium.io/documentation//reference/transformations/event-flattening.html[New record state SMT], which does exactly this - extracts only new state from the Debezium record.
The configuration of the PostgreSQL source connector now looks like this:

[source, json, options="nowrap"]
----
{
    "name": "postgres-source",
    "config": {
        "connector.class": "io.debezium.connector.postgresql.PostgresConnector",
        "tasks.max": "1",
        "database.hostname": "postgres",
        "database.port": "5432",
        "database.user": "postgres",
        "database.password": "postgres",
        "database.dbname" : "postgres",
        "topic.prefix": "polls",
        "table.include.list": "public.optionentity, public.pollentity, public.pollentity_optionentity, public.pollentity_voteentity, public.voteentity",
        "transforms": "unwrap",
        "transforms.unwrap.type": "io.debezium.transforms.ExtractNewRecordState",
        "transforms.unwrap.add.fields": "table"
    }
}
----

In the SMT configuration we also add additional field - the name of the table which the record comes from.

> Another difference from the Debezium sink connectors is that single sink connector ingest data only to a single table.

Reviewer: Are you sure this is the case? Reading the docs https://questdb.com/docs/third-party-tools/kafka/#configuration-options I'd say that by default there is topic to table mapping and everything can be routed into a single table but I might be wrong (also there is https://github.com/questdb/kafka-questdb-connector/blob/main/connector/src/main/java/io/questdb/kafka/Templating.java#L17).

Author: Thanks a lot for spotting this, it's indeed as you wrote (it was strange to me that it wouldn't allow multiple table, but I somehow miss it:-(). Now it's more simple and much more nicer.

        "tasks.max": "5",
        "topics": "polls.public.voteentity",
        "table": "votes",
        "client.conf.string": "http::addr=questdb:9000;",
        "timestamp.field.name": "ts",
        "value.converter": "org.apache.kafka.connect.json.JsonConverter",
        "value.converter.schemas.enable": true,
        "include.key": false,
        "key.converter": "org.apache.kafka.connect.storage.StringConverter"
    }
}
----

This configuration ensures that poll-related data and vote data are written into separate QuestDB tables.
Even though this setup requires slightly more configuration compared to the JDBC sink, it remains straightforward and easy to configure.

== Application point of view

We have shown several ways how to replicate data from one database instance to the other as a main building block for implementing CQRS pattern.
No matter which setup you choose, the end result from the user perspective should be always the same and you should see something like this when looking on the poll results:

++++
<div class="imageblock centered-image">
	<img src="/assets/images/2025-10-14-cqrs/poll-results.png" class="responsive-image" alt="Demo application ">
</div>
++++

> The results plot should be automatically update as the votes are castes by the voters.

Reviewer: ```suggestion
The results plot should be automatically updated as the votes are casted by the voters.
```

=== Summary

CQRS design pattern is a powerful pattern which suggests to split the application into part responsible for manipulating with the database data and part responsible for reading this data.
This separation allows us to use different databases for each microservice.
Independent databases allows independent scaling and tuning adjusted exactly to their needs and workload of given microservice.
This segregation also allows different data models for each microservice allowing fine tuning of the models, but also independent evolution of each microservice.
Other benefits of implementing this pattern involves increased security as only microservice which push data into database needs to have write access, better clarity and simplicity of the code and more.

In this blog post, besides introducing the pattern itself, we concentrate on the main obstacle in implementation of this pattern - how to replicate the data between the databases.
We discussed several possible architectures and described ways how to ensure data replication in each setup.

> We started with the simple case when the replication is ensured by database itself, continued with replication between two different relational databases and concluded with setup where the relational database is replicated in to special-purpose database.

Reviewer: ```suggestion
We started with the simple case when the replication is ensured by database itself, continued with replication between two different relational databases and concluded with setup where the relational database is replicated into special-purpose database.
```
